File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py

from collections import deque
def dfs(x):
    visited[x] = True
    print(x, end=" ")
    for node in graph[x]:
        if not visited[node]:
            dfs(node)


def bfs(x):
    queue = deque([x])
    visited[x] = True
    while queue:
        v = queue.popleft()
        print(v, end=" ")
        for i in graph[v]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True


n, m, v = map(int, input().split())
graph = [[] for _ in range(n + 1)]
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)
    graph[a].sort()
    graph[b].sort()


visited = [False] * (n + 1)
dfs(v)
print()

visited = [False] * (n + 1)
bfs(v)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

val sb = StringBuilder()

fun dfs(graph: Array<MutableList<Int>>, start: Int, visited: BooleanArray) {
    visited[start] = true
    sb.append("$start ")

    for (next in graph[start]) {
        if (!visited[next]) {
            dfs(graph, next, visited)
        }
    }
}

fun bfs(graph: Array<MutableList<Int>>, start: Int, visited: BooleanArray) {
    val queue: Queue<Int> = LinkedList()
    queue.add(start)
    visited[start] = true

    while (queue.isNotEmpty()) {
        val node = queue.poll()
        sb.append("$node ")

        for (next in graph[node]) {
            if (!visited[next]) {
                queue.add(next)
                visited[next] = true
            }
        }
    }
}

fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(reader.readLine())
    val n = st.nextToken().toInt()
    val m = st.nextToken().toInt()
    val v = st.nextToken().toInt()

    val graph = Array(n + 1) { mutableListOf<Int>() }
    repeat(m) {
        val (a, b) = reader.readLine().split(" ").map { it.toInt() }
        graph[a].add(b)
        graph[b].add(a)
    }
    for (i in 1..n) {
        graph[i].sort() // 정점이 여러 개인 경우에는 정점 번호가 작은 것을 먼저 방문
    }

    var visited = BooleanArray(n + 1) { false }
    dfs(graph, v, visited)
    sb.append("\n")

    visited = BooleanArray(n + 1) { false }
    bfs(graph, v, visited)

    println(sb)
}